### PR TITLE
METRON-1844: Allow for LDAP to be used for authentication and roles

### PR DESCRIPTION
**Pull request by justinleet**
## Contributor Comments
I went through the defunct Knox SSO feature branch and pulled out the LDAP bits. I made sure @simonellistonball is on the first commit to maintain attribution.  When we merge this, we'll need to make sure that's properly maintained.

What this does is setup the set of LDAP configs, while maintaining backwards compatibility with the JDBC stuff by default.

I'd say the main objection I'd expect would be around unit/integration testing. Right now there's not anything in this PR, although we could probably pull in the EmbeddedLDAP from the branch and make sure the role endpoint works with it for integration.

I have run this up on full dev and tested against an actual OpenLDAP instance using https://github.com/osixia/docker-openldap. Instructions will be provided in comment to duplicate this testing.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [ ] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by justinleet** [comment]
I updated the Ambari config to use a toggle On/Off for the LDAP functionality.  The adjustment to the spring profiles takes place in the background, so the user doesn't have to worry about it.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py [hunk 1]
```diff
@@ -267,6 +274,21 @@
     if 'solr-config-env' in config['configurations']:
         solr_principal_name = solr_principal_name.replace('_HOST', hostname_lowercase)
 
+# LDAP
+metron_ldap_url = config['configurations']['metron-security-env']['metron.ldap.url']
+metron_ldap_userdn = config['configurations']['metron-security-env']['metron.ldap.bind.dn']
+metron_ldap_password = config['configurations']['metron-security-env']['metron.ldap.bind.password']
+metron_ldap_user_pattern = config['configurations']['metron-security-env']['metron.ldap.user.dnpattern']
+metron_ldap_user_password = config['configurationsmetron_spring_profiles_active']['metron-security-env']['metron.ldap.user.password']
```
Comment (anandsubbu) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: Looks like a mispaste. `metron_spring_profiles_active` should be removed.

Comment (justinleet) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: Good catch. Fixed in latest commit

**Comment by justinleet** [comment]
Updated the instructions above with corrected LDAP URL

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-rest-env.xml [hunk 2]
```diff
@@ -35,32 +35,29 @@
...
     <property>
         <name>metron_spring_profiles_active</name>
-        <description>Active Spring profiles</description>
+        <description>Active Spring profiles. 'ldap' is used to enable authentication via LDAP.</description>
         <display-name>Active Spring profiles</display-name>
-        <value/>
-        <value-attributes>
-            <empty-value-valid>true</empty-value-valid>
-        </value-attributes>
+        <value>jdbc</value>
```
Comment (merrimanr) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-rest-env.xml]: Are we using a "jdbc" profile anywhere?  From what I can tell jdbc authentication is used by default but we don't actually use a profile for that.

Comment (justinleet) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-rest-env.xml]: good catch. I thought I'd rolled that back.  I'll udpate

**Comment by merrimanr** [comment]
I tested this in full dev using the instructions in the README and with the Knox sample ldap.  Both worked fine.  I also tested using legacy JDBC authentication and it continued to work.  Nice job!  +1

**Comment by anandsubbu** [comment]
My +1 as well. 

I spun up a multi-node cluster and set this up using the demo LDAP service provided by Knox. I was able to authenticate using the demo users for all the Swagger, Alerts and Management UI interfaces. Nicely done, @justinleet !

**Pull request closed without merging**